src/lib/homeobject_impl.hpp

```diff
@@ -46,33 +56,34 @@ class HomeObjectImpl : public HomeObject,
                        public std::enable_shared_from_this< HomeObjectImpl > {
 
     /// Implementation defines these
-    virtual ShardManager::Result< ShardInfo > _create_shard(pg_id, uint64_t size_bytes) = 0;
-    virtual ShardManager::Result< ShardInfo > _seal_shard(shard_id) = 0;
+    virtual ShardManager::Result< ShardInfo > _create_shard(pg_id_t, uint64_t size_bytes) = 0;
+    virtual ShardManager::Result< ShardInfo > _seal_shard(shard_id_t) = 0;
 
-    virtual BlobManager::Result< blob_id > _put_blob(ShardInfo const&, Blob&&) = 0;
-    virtual BlobManager::Result< Blob > _get_blob(ShardInfo const&, blob_id) const = 0;
-    virtual BlobManager::NullResult _del_blob(ShardInfo const&, blob_id) = 0;
+    virtual BlobManager::Result< blob_id_t > _put_blob(ShardInfo const&, Blob&&) = 0;
+    virtual BlobManager::Result< Blob > _get_blob(ShardInfo const&, blob_id_t) const = 0;
+    virtual BlobManager::NullResult _del_blob(ShardInfo const&, blob_id_t) = 0;
     ///
-    folly::Future< ShardManager::Result< Shard > > _get_shard(shard_id id) const;
+    folly::Future< ShardManager::Result< Shard > > _get_shard(shard_id_t id) const;
     auto _defer() const { return folly::makeSemiFuture().via(folly::getGlobalCPUExecutor()); }
 
+    virtual PGManager::NullAsyncResult _create_pg(PGInfo&& pg_info, std::set< std::string, std::less<> > peers) = 0;
+    virtual PGManager::NullAsyncResult _replace_member(pg_id_t id, peer_id_t const& old_member,
+                                                       PGMember const& new_member) = 0;
+
 protected:
     std::mutex _repl_lock;
-    std::shared_ptr< home_replication::ReplicationService > _repl_svc;
-    // std::shared_ptr<homestore::ReplicationService> _repl_svc;
-    peer_id _our_id;
+    peer_id_t _our_id;
 
     /// Our SvcId retrieval and SvcId->IP mapping
     std::weak_ptr< HomeObjectApplication > _application;
 
     ///
     mutable std::shared_mutex _pg_lock;
-    std::map< pg_id, PG > _pg_map;
+    std::map< pg_id_t, std::shared_ptr< PG > > _pg_map;
```

Comment:
I'm not following why this has to be a shared_ptr? And shouldn't we use `shared` if we aliased it?

Reply:
Here is the deal: I retained it as shared_ptr, because adding a base class into map and accessing them as derived class doesn't work with or without polymorphism. It is safe to add it as shared_ptr. We can obviously remove the entire concept of this, if we have different map (one for memory and other for homeobject). 

Another `TODO` is that we are taking global lock `_pg_lock` instead of per pg lock. I didn't change that part now, but we must change that.